src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,188 @@
+		}
+	}
+
+	public async publish(ctx: LogWriteContext, logs: DeviceLog[]): Promise<any> {
+		return requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/push`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'POST',
+			json: {
+				streams: logs.map((log) => this.toLokiStream(ctx, log), this),
+			},
+		});
+	}
+
+	public subscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		if (!this.webSockets.has(key)) {
+			const tailUri =
+				`ws://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/tail` +
+				`?query=${encodeURIComponent(this.getDeviceQuery(ctx))}&start=${String(
+					Date.now(),
+				)}000000`;
+			const ws = new WebSocket(tailUri, {
+				headers: { 'X-Scope-OrgID': ctx.belongs_to__application },
+			});
+			ws.on('message', (data: string) => {
+				this.fromLokiStreamsToDeviceLogs(JSON.parse(data).streams).map(
+					(log: DeviceLog) => {
+						this.subscriptions.emit(key, log);
+					},
+				);
+			});
+			ws.on('close', () => {
+				this.webSockets.delete(key);
+			});
+			this.webSockets.set(key, ws);
+		}
+		this.subscriptions.on(key, subscription);
+	}
+
+	public unsubscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		this.subscriptions.removeListener(key, subscription);
+		if (!this.webSockets.has(key)) {
+			const ws = this.webSockets.get(key);
+			if (ws) ws.close();
+		}
+	}
+
+	private getDeviceQuery(ctx: LogContext) {
+		return `{device_id="${ctx.id}"}`;
+	}
+
+	private getKey(ctx: LogContext, suffix = 'logs') {
+		return `device:${ctx.id}:${suffix}`;
+	}
+
+	private fromLokiStreamsToDeviceLogs(streams: LokiStream[]): DeviceLog[] {
+		return _.concat(
+			[],
+			...streams.map((stream: LokiStream) => this.fromLokiStream(stream)),
+		);
+	}
+
+	private fromLokiStream(stream: LokiStream): DeviceLog[] {
+		try {
+			return stream.values.map((value) => {
+				const [, message] = value;
+				const log = JSON.parse(message);
+				if (log.version !== VERSION) {
+					throw new Error(
+						`Invalid Redis serialization version: ${JSON.stringify(log)}`,
+					);
+				}
+				if (log.serviceId === null) {
+					delete log.serviceId;
+				}
+				delete log.version;
+				return log as DeviceLog;
+			});
+		} catch (err) {
+			captureException(err, `Failed to deserialize a Loki log: ${stream}`);
+			return [];
+		}
+	}
+
+	private toLokiStream(ctx: LogContext, log: DeviceLog): LokiStream {
+		try {
+			skhema.validate(schema, log);
+			return {
+				stream: {
+					device_id: String(ctx.id), // write only device id, to reduce cardinality do not also write device uuid
+					service_id: String(log.serviceId),
+					is_system: String(log.isSystem),
+					is_std_err: String(log.isStdErr),
+				},
+				values: [
+					[
+						`${log.timestamp.toString()}000000`, // convert timestamp to nanosecond string
+						JSON.stringify({ version: VERSION, ...log }), // store log as JSON string to retain any meta not writen as label
```

Comment:
I think doing
```
log.version = VERSION
JSON.stringify(log)
```
will have much better performance

Reply:
TS complains `TS2339: Property 'version' does not exist on type 'DeviceLog'.` do you recommend that I use `// @ts-ignore`?

Reply:
You should be able to do
```ts
	private toLokiStream(ctx: LogContext, log: DeviceLog & { version?: number}): LokiStream {
```
and type it properly